### ISSUE TITLE
Text blanks, fails to read from file

### ISSUE THREAD

**Issue by Meerkov** [obsproject/obs-studio#3011]
## Platform
Windows 10 
OBS Studio version: 25.0.1 64 bit

## Expected Behavior
1) Create a Text Source with "Read File" enabled.
2) Switch to a scene where the Text Source is deleted
3) Update the text file manually in another program (or use a script in a separate process).
4) Switch back to the first scene
5) Repeat step 2-4 another 10x

Result: The text updates every time

## Current Behavior
Sometimes the TextSource will become blank even though the file is not empty.
The TextSource may sometimes recover if content is written to the file again, but sometimes will not.

## Steps to Reproduce
1) run this in OBS. Make sure to have a text source with "Read File" on one scene, but deleted on the other.
```lua
obs = obslua

function change_scene(sceneName)
	source = obs.obs_get_source_by_name(sceneName)
	if source ~= nil then
		obs.obs_frontend_set_current_scene(source)
		obs.obs_source_release(source)
	else
		print("Can't find scene: "..sceneName)
	end
end

function display_timer_tick()
	if(state == "DisplayText") then
                state == "HideTextDuringUpdate"
		change_scene("s1")
        else
                state == "DisplayText"
                change_scene("s2")
        end
end

function script_load(settings)
	print("Loaded Script.")
        obs.timer_add(display_timer_tick, 10000)
end
```
2) Update the textfile while the text is hidden.
3) Watch to see if the text appears as expected.
4) if it appears, edit the textfile and save again when it disappears. (I used a separate script to synchronize the file-write with OBS's scene switching, but you should be able to do this manually instead).
5) Repeat until it shows up blank.
6) Stop editing the file and wait to see it probably doesn't read the file until you edit it again.
If it doesn't work, try "refreshing" the lua script in OBS.

## Additional information

It might be something related to the source getting disabled during scene switching, which the documentation acknowledges "is confusing to users because it's not surfaced in the UI and sources will fail to render even though they appear to be marked visible". But I would have no idea what might cause that to happen here.

**Comment by Meerkov** [comment]
It triggered for me again. This time with manually timed file-reads via a script.

```lua

function update_text(sourcename, value )
  print("Updating text for "..sourcename)
  text=obs.obs_get_source_by_name(sourcename)
  if(text ~= nil) then
	newRef=obs.obs_data_create_from_json('{"text":"'..value..'"}')
	ref=obs.obs_save_source(text)
	obs.obs_source_update(text,newRef)
	-- print(obs.obs_data_get_json(ref)) -- debugging
	obs.obs_data_release(ref)
	obs.obs_data_release(newRef)
  end
end

function read_file(filename)
		file = io.open(filename,"r")
		io.input(file)
		tmp = io.read()
		io.close()
	return tmp
end

function display_timer_tick()
		update_text("Name", read_file("C:/text_file.txt"))
end

function script_load(settings)
	print("Loaded Script.")
    obs.timer_add(display_timer_tick, 1000)
end
```

Run this script, and have a separate script that writes a short (10 byte) string to text_file every 3 minutes or so. Eventually it will fail, and error output will say that it can't concatenate a NIL. This is because `value` is nil.

However, it doesn't recover, and every second it will output the same error message.

**Comment by notr1ch** [comment]
Are you performing atomic updates of the text file? If not, OBS may be reading the file as soon as it's open for writing but before the contents are written.

**Comment by YorVeX** [comment]
> Are you performing atomic updates of the text file? If not, OBS may be reading the file as soon as it's open for writing but before the contents are written.

At least for #3275 I am pretty sure that's what happens. IMHO in these cases OBS should retry to read the file later. Otherwise it means that you can create situations where the OBS source does not reflect the contents of the file anymore, but that is the whole idea of that mechanism and I want to be able to trust it.
For performance reasons OBS should maybe not try too aggressively to re-access the file (it might just be continuously locked and never readable) but once per second shouldn't do too much harm.

On any file change just remember some kind of "dirty" flag for the source and while that flag is set try every second to read it and update the source. Reset the "dirty" flag as soon as the file read was successful.

**Comment by YorVeX** [comment]
I just changed my program to write to a different file first, then delete the original file and rename the new one into that place so that this all happens faster. But even with this workaround the problem still occurred, I am looking at an image shown by the image source in OBS that is differing from what is in the image file right now.

Even with my change it is not a completely atomic operation, since file deletion and move are two different operations. But the interesting thing is: I am not looking at an empty image source, just an outdated one. So the OBS update didn't occur between the delete and move. I'd rather assume OBS tried to read the file during the move operation, failed and therefore kept the image state from the last working file read. It seems that not even a file move operation is fast enough to avoid this problem.

Either way, for me this means one cannot even easily avoid this problem with control over the program/code that writes the file (which is not the case for everyone anyway). It must really be fixed on OBS side.

**Comment by RytoEX** [comment]
Can someone provide reliable reproduction steps for this? From a review of our last findings, we had trouble reproducing this reliably to further investigate.

**Comment by YorVeX** [comment]
> Can someone provide reliable reproduction steps for this? From a review of our last findings, we had trouble reproducing this reliably to further investigate.

Depends on at which part you are looking at.
> The TextSource may sometimes recover if content is written to the file again, but sometimes will not.

Regarding the "but sometimes will not" part: I know I certainly did have the problem that sometimes the image source would never recover anymore after a read failed, not even if the image file was overwritten again with new data. **This** problem I cannot reproduce anymore, just tried with a script that would keep on overwriting the underlying file of an image source every 3 seconds for 3 hours straight. It did cause the occasional load error as expected when the load attempt happened to be performed at the exact moment of the copy operation:
```
23:30:10.465: [image_source: 'Example Image'] loading texture 'X:/Programs/OBS Studio/bin/64bit/example.jpg'
23:30:10.466: Failed to open file 'X:/Programs/OBS Studio/bin/64bit/example.jpg': Permission denied
23:30:10.466: WIC: Failed to create IWICBitmapDecoder from file: X:/Programs/OBS Studio/bin/64bit/example.jpg
23:30:10.466: gs_image_file_init_internal: Failed to load file 'X:/Programs/OBS Studio/bin/64bit/example.jpg'
23:30:10.466: [image_source: 'Example Image'] failed to load texture 'X:/Programs/OBS Studio/bin/64bit/example.jpg'
```
The next overwrite would then always make the source recover (but even back then it only happened quite rarely, so maybe it was just "bad luck").

However, the first problem that after a single failed read attempt the source only recovers if the file is updated once again - **that** definitely is still there and IMHO is not a bug that needs reproducing, it's clearly visible from the code that it's currently designed like this. You can see in [line 54](https://github.com/obsproject/obs-studio/blob/29.0.2/plugins/image-source/image-source.c#L54) that file_timestamp is updated before the file is actually read in line 55. Now if reading the file fails (as indicated by the error in line 66 that I can also see in my logs) the field file_timestamp that controls whether OBS tries to read the file again signalizes "this file was already read", hence OBS never tries again (see [lines 208](https://github.com/obsproject/obs-studio/blob/29.0.2/plugins/image-source/image-source.c#L208) to 212 for the relevant part of the code).

Those examples are from the image source, but I checked and the code works the same for the text source.

That it's designed like that doesn't mean it's good. It doesn't make sense to me that the source would attempt every second to load a new file, but as soon as one read failed it stops these attempts, although technically for the source the file is still "new", since it didn't successfully load it so far. From my point of view it should just keep on retrying every second and only update the internal timestamp when loading succeeded.

**Comment by YorVeX** [comment]
If you do want to try and reproduce it nevertheless I can only help for Windows, here's a PowerShell script that will copy example1.png to example.png (which you are supposed to have an image source pointed to) so that it would update the file timestamp and trigger a reload in OBS, and then it immediately locks example.png for 1.5 seconds to make sure the file is definitely locked during the next attempt of OBS to access the file.

```powershell
$source = "example1.png"
$target = "example.png"

    Copy-Item $source $target -Force
    Write-Host "Locking $target"
    $fileStream = New-Object IO.FileStream($target, [System.IO.FileMode]::Open, [System.IO.FileAccess]::ReadWrite, [System.IO.FileShare]::None)
    Start-Sleep -Milliseconds 1500
    $fileStream.Close()
    Write-Host "Lock released"

```

**Comment by Fenrirthviti** [comment]
Closing as stale. If this is still an issue please open a new issue with updated reproduction steps on the latest version of OBS.

From my understanding of this based on the comments, this might be a process workflow issue more than a bug in OBS. Updated, clear reproduction steps in a new issues on the latest version would be preferred if this is still a problem.

**Comment by Meerkov** [comment]
The ultimate purpose of this was to have a python script running in parallel, sending signals or text updates to the UI through files. This is because I was running some simulation that needed to calculate a "winner", and then set a text file to the name of the winner of a match.

===

As far as I remember, it's the exact same issue as described by the image issue, but for text files.

The setup I had:
1. OBS is running that script.
2. Every 10 seconds it switches from scene 1 to scene 2, or viceversa.
3. One of the scenes has the a text element that is pointing to a simple text file.
4. While on the scene where you can't see it, you update the text file.
5. When the scene switches, you may or may not see the text actually update.

From what I recall, this had to do with OBS's "last updated" timestamps.

There is probably also a confounding issue which is I'm assuming OBS is only checking last update time to the nearest second. So it may "check the file" at ::10 seconds, the file is updated at ::10.9 seconds, but OBS sees the same timestamp as it's last check and ignores the update.
